app/models/lock.rb

```diff
@@ -51,9 +58,54 @@ def delete_in=(seconds)
     self.delete_at = seconds.present? ? Time.now + seconds.to_i : nil
   end
 
+  def affected
+    if resource_type == "Stage"
+      "stage"
+    elsif resource
+      resource.name
+    else
+      "ALL STAGES"
+    end
+  end
+
+  # normally there are very few locks, so we grab them all and filter down to avoid lookups
+  # sorted by priority(warning) and specificity(type)
+  def self.for_resource(resource)
+    matching_locks = all_cached.select do |lock|
+      lock.global? ||
+        lock.resource_equal(resource) ||
+        (resource.class.name == "Stage" && resource.environments.any? { |e| lock.resource_equal(e) })
```

Comment:
hmm yes ... but it would be `if lock.resource_type == "Environment" && resource.is_a?(Stage) && resource.environments { foo }` ... I'll add that :)